llvm/lib/Target/X86/X86ISelLowering.cpp

```diff
@@ -23489,6 +23488,23 @@ static SDValue EmitCmp(SDValue Op0, SDValue Op1, X86::CondCode X86CC,
     Op1 = DAG.getNode(ISD::TRUNCATE, dl, CmpVT, Op1);
   }
 
+  // Try to shrink signed i64 compares if the input has enough one bits.
+  // Or the input is sign extended from a 32-bit value.
+  // TODO: Should we peek through freeze?
+  // TODO: Is SIGN_EXTEND_INREG needed here?
+  if (CmpVT == MVT::i64 && isX86CCSigned(X86CC) &&
+      Op0.hasOneUse() && // Hacky way to not break CSE opportunities with sub.
+      (DAG.ComputeNumSignBits(Op1) > 32 ||
+       Op1.getOpcode() == ISD::SIGN_EXTEND ||
```

Comment:
Why are we checking for specific opcodes? Aren't they already handled by ComputeNumSignBits?